internal/eva/role.go

```diff
@@ -0,0 +1,168 @@
+package eva
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+)
+
+const (
+	createRolePath = "/api/core/management/CreateRole"
+	getRolePath    = "/api/core/management/GetRole"
+	updateRolePath = "/api/core/management/UpdateRole"
+	deleteRolePath = "/api/core/management/DeleteRole"
+)
+
+type CreateRoleRequest struct {
+	Name     string `json:"Name"`
+	UserType int64  `json:"UserType,omitempty"`
+	Code     string `json:"Code,omitempty"`
+}
+type CreateRoleResponse struct {
+	ID int64 `json:"ID"`
+}
+
+func (c *Client) CreateRole(ctx context.Context, req CreateRoleRequest) (*CreateRoleResponse, error) {
+	resp, err := c.restClient.R().
+		SetBody(req).
+		Post(createRolePath)
+
+	if err != nil {
+		tflog.Error(ctx, "An network error ocurred.", err)
+
+		return nil, err
+	}
+
+	if resp.StatusCode() != 200 {
+		tflog.Info(ctx, "Request failed", "Status code", resp.StatusCode(), "body", resp.String())
+
+		return nil, errors.New(fmt.Sprintf("Request failed with error: %s", resp.String()))
+	}
+
+	var jsonResp CreateRoleResponse
+	if err := json.Unmarshal([]byte(resp.Body()), &jsonResp); err != nil {
+		return nil, errors.New(fmt.Sprintf("Response could not be parsed. Received: %s", resp.String()))
+	}
+
+	return &jsonResp, nil
+}
+
+type GetRoleRequest struct {
+	ID int64 `json:"ID"`
+}
+
+type Role struct {
+	Name                  string                    `json:"Name"`
+	UserType              int64                     `json:"UserType,omitempty"`
+	Code                  string                    `json:"Code,omitempty"`
+	ScopedFunctionalities []RoleScopedFunctionality `json:"ScopedFunctionalities"`
```

Comment:
I created this struct `RoleScopedFunctionality` which is the same as `ScopedFunctionality` inside `role_permissions.go`, is still a good practise to re-use it ? or do I have to always create a new struct ? @mad-it

Reply:
I would say that even if the fields are the same I would keep them separate for now because we don't know if the underlying API re-uses them or if they are different. If they are different we run the risk of them changing under the hood and that would break two of our dependencies